### ISSUE TITLE
Generic.PHP.DisallowShortOpenTag doesn't catch `<?`

### ISSUE THREAD

**Issue by Flimm** [squizlabs/PHP_CodeSniffer#1571]
Let's take this example PHP file:

```
<?

// This is a test
```

If I run `phpcs` against it, I get no errors or warnings:

```
$ phpcs --sniffs=Generic.PHP.DisallowShortOpenTag example.php
$
```

There are several code standards that advise against the short open tag `<?`, it seems that this sniffer is only checking for `<?=` and not for `<?`. `DisallowShortOpenTag` should definitely check for `<?` short opening tags. It could also autocorrect them to `<?php`.

(Arguably, only the `<?` short open tag is problematic, and `<?=` isn't, but that's another discussion for another issue. See [this Stack Exchange answer](https://softwareengineering.stackexchange.com/a/311257/3025) and [this one](https://softwareengineering.stackexchange.com/a/151694/3025).)

**Comment by Flimm** [comment]
Weirdly, when I don't pass a `--sniffs=Generic.PHP.DisallowShortOpenTag` argument, I get this warning:

```
$ phpcs example.php

FILE: example.php
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 1 | WARNING | No PHP code was found in this file and short open tags are not allowed by this install of PHP. This file may be using short open tags but PHP does not allow them.
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

If however, I make sure to include both `<?php` and `<?` opening tags, I don't get any errors about the short code. For instance, with this file:

```
<?php

echo "one";

?>
<?

echo "two";
```

I get this output:

```
$ phpcs example.php

FILE: example.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 2 | ERROR | Missing file doc comment
----------------------------------------------------------------------
```

So this is still a valid bug report.

**Comment by gsherwood** [comment]
This is because your PHP version has short open tags set to off. So when PHP tokenizes that file, it doesn't find any PHP code inside those <? ?> tags, so PHPCS doesn't see anything in there either. It would be tokenized as `T_INLINE_HTML`.

When PHPCS doesnt find any PHP in the file at all, it gives a warning about short open tags because this is a common reason why a PHP file contains *only* T_INLINE_HTML.

When I run your second code sample using a PHP version that enables short open tags, I get this:
```
FILE: temp.php
----------------------------------------------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------------------------------------------------------
 2 | ERROR | [ ] Missing file doc comment (PEAR.Commenting.FileComment.Missing)
 6 | ERROR | [x] Short PHP opening tag used; expected "<?php" but found "<?" (Generic.PHP.DisallowShortOpenTag.Found)
----------------------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------------------------------------------------------
```
So this sniff is working fine. It's just that PHP can't see the code inside your short open tags, so that is hiding it from PHPCS. Yes, PHPCS could inspect every single T_INLINE_HTML token and see if there is a <? string in there, and then try and figure out if that is actually for PHP code and not some other scripting language, but it's not something I think the sniff should spend time doing.

Hopefully that clears the behaviour up for you.

**Comment by Flimm** [comment]
Thanks @gsherwood , that explanation makes sense.

We could improve this warning:

```
 1 | WARNING | No PHP code was found in this file and short open tags are not allowed by this install of PHP. This file may be using short open tags but PHP does not allow them.
```

We could make this warning only trigger when the file (the T_INLINE_HTML token, I guess) contains the string `<?`. If the file doesn't contain `<?` anywhere, there is no need for this to trigger. We could also make it an error and part of `Generic.PHP.DisallowShortOpenTag`.

**Comment by jrfnl** [comment]
> PHPCS could inspect every single T_INLINE_HTML token and see if there is a <? string in there, and then try and figure out if that is actually for PHP code and not some other scripting language

The sniff actually already does this, but to not create lots of false positives, it will only report on the open tag in that case if it also finds a close tag.

So:
* ini_value `short_open_tags` set to `on` means you will receive an `error` when using short open tags.
* ini_value `short_open_tags` set to `off` means you may receive a `warning` about possible use of short open tags, but only when a complete set of short open tags are found.
* With ini_value `short_open_tags` set to `off`, the sniff will not be triggered if no `<?php` tags are found at all in the file as the file is in that not considered PHP code and will trigger the `Internal.NoPHPCode` warning mentioned before.

Try this code to see it in action:
```php
<?php
// Some code

?>
<img src="<? echo $src; ?>"/>
```

**Comment by gsherwood** [comment]
> The sniff actually already does this, but to not create lots of false positives, it will only report on the open tag in that case if it also finds a close tag.

I forgot you did up a PR for this. Probably why I forgot how it works.

So @Flimm does that explain things then?

**Comment by Flimm** [comment]
Thanks for the explanation.

The current behaviour seems great, but could be better. Consider the case of an empty file. When `short_open_tag` is `on`, no warnings are generated, which is as I hoped and expected:

```
$ printf "" > empty.php
$ php -i | grep short_open_tag
short_open_tag => On => On
$ phpcs empty.php
$
```

When `short_open_tag` is `off`, however, this warning is generated. I would prefer that the code were smart enough to notice that the T_SINGLE_HTML token is empty and doesn't contain `<?`, and so therefore there is no need for this warning:

```
$ php -i | grep short_open_tag
short_open_tag => Off => Off
$ phpcs empty.php

FILE: empty.php
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 1 | WARNING | No PHP code was found in this file and short open tags are not allowed by this install of PHP. This file may be using short open tags but PHP does not allow them.
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

I'm no longer confused as to why PHPCS behaves as it does, but it feels that it could be improved to be more useful.

I'm closing because my questions have been answered, if someone wants to improve Internal.NoPHPCode, I would welcome that.
